source/MaterialXCore/Document.cpp

```diff
@@ -782,12 +830,170 @@ void Document::upgradeVersion()
 
         minorVersion = 37;
     }
+    if (majorVersion == desiredMajorVersion &&
+        minorVersion == desiredMinorVersion)
+    {
+        setVersionString(makeVersionString(majorVersion, minorVersion));
+        return;
+    }
 
-    if (majorVersion == MATERIALX_MAJOR_VERSION &&
-        minorVersion == MATERIALX_MINOR_VERSION)
+    if (majorVersion == 1 && minorVersion == 37)
     {
-        setVersionString(DOCUMENT_VERSION_STRING);
+        // Convert material Elements to Nodes
+        convertMaterialsToNode(true);
+        minorVersion = 38;
+    }
+
+    if (majorVersion >= MATERIALX_MAJOR_VERSION &&
+        minorVersion >= MATERIALX_MINOR_VERSION)
+    {
+        setVersionString(makeVersionString(majorVersion, minorVersion));
+    }
+}
+
+bool Document::convertMaterialsToNode(bool replaceNodes)
+{
+    bool modified = false;
+
+    vector<MaterialPtr> materials = getMaterials();
+    if (!materials.empty())
+    {
+        for (auto m : materials)
+        {
+            // See if a node of this name has already been created.
+            // Should not occur otherwise there are duplicate existing
+            // Material elements.
+            string materialName = m->getName();
+            if (getNode(materialName))
+            {
+                continue;
+            }
+
+            // Create a temporary name for the material element 
+            // so the new node can reuse the existing name.
+            if (replaceNodes)
+            {
+                string validName = createValidChildName(materialName + "1");
+                m->setName(validName);
+            }
+            else
+            {
+                materialName = createValidChildName(materialName);
+            }
+
+            // Create a new material node
+            NodePtr materialNode = nullptr;
+
+            ShaderRefPtr sr;
+            // Only include the shader refs explicitly specified on the material instance
+            vector<ShaderRefPtr> srs = m->getShaderRefs();
+            for (size_t i = 0; i < srs.size(); i++)
+            {
+                sr = srs[i];
+
+                // See if shader has been created already.
+                // Should be unlikely to occur as the shaderref is a uniquely named
+                // child of a uniquely named material element, but the two combined
+                // may have been used for another node instance which not shader node.
+                string shaderNodeName = m->getName() + "_" + sr->getName();
+                NodePtr existingShaderNode = getNode(shaderNodeName);
+                if (existingShaderNode)
+                {
+                    const string& existingType = existingShaderNode->getType();
+                    if (existingType == VOLUME_SHADER_TYPE_STRING ||
+                        existingType == SURFACE_SHADER_TYPE_STRING ||
+                        existingType == DISPLACEMENT_SHADER_TYPE_STRING)
+                    {
+                        continue;
```

Comment:
once again, is just continuing a good solution?

Reply:
I think maybe this can throw an exception.